lib/sms/sms_deliver.c

```diff
@@ -0,0 +1,808 @@
+	 *  meaning. Otherwise bits 1 to 0 are reserved.
+	 */
+	uint8_t presence_of_class:1;
+	/** If set to 0, indicates the text is uncompressed.
+	 *  Otherwise it's compressed.
+	 */
+	uint8_t compressed:1;
+};
+
+/**
+ * @brief SMS-DELIVER PDU fields specified in 3GPP TS 23.040 chapter 9.2.2.1
+ */
+struct pdu_deliver_data {
+	struct pdu_deliver_header   field_header; /** First byte of header */
+	struct sms_address          field_oa;  /** TP-Originating-Address */
+	uint8_t                     field_pid; /** TP-Protocol-Identifier */
+	struct pdu_dcs_field        field_dcs; /** TP-Data-Coding-Scheme */
+	struct sms_time             timestamp; /** TP-Service-Centre-Time-Stamp */
+	uint8_t                     field_udl; /** TP-User-Data-Length */
+	uint8_t                     field_udhl; /** User Data Header Length */
+	struct sms_udh_app_port     field_udh_app_port; /** Port addressing */
+	struct sms_udh_concatenated field_udh_concatenated; /** Concatenation */
+	uint8_t                     field_ud[140]; /** TP-User-Data */
+};
+
+/**
+ * @brief Swap upper and lower 4 bits between each other.
+ */
+static uint8_t swap_nibbles(uint8_t value)
+{
+	return ((value&0x0f)<<4) | ((value&0xf0)>>4);
+}
+
+/**
+ * @brief Converts an octet having two semi-octets into a decimal.
+ *
+ * @details Semi-octet representation is explained in 3GPP TS 23.040 Section 9.1.2.3.
+ * An octet has semi-octets in the following order:
+ *   semi-octet-digit2, semi-octet-digit1
+ * Octet for decimal number '21' is hence represented as semi-octet bits:
+ *   00010010
+ * This function is needed in timestamp (TP SCTS) conversion that is specified
+ * in 3GPP TS 23.040 Section 9.2.3.11.
+ *
+ * @param[in] value Octet to be converted.
+ *
+ * @return Decimal value.
+ */
+static uint8_t semioctet_to_dec(uint8_t value)
+{
+	/* 4 LSBs represent decimal that should be multiplied by 10. */
+	/* 4 MSBs represent decimal that should be add as is. */
+	return ((value & 0xf0) >> 4) + ((value & 0x0f) * 10);
+}
+
+/**
+ * @brief Convert phone number into string format.
+ *
+ * @param[in] number Number in semi-octet representation.
+ * @param[in] number_length Number length.
+ * @param[out] str_number Output buffer where number is stored. Size shall be at minimum twice the
+ *                        number length rounded up.
+ */
+static void convert_number_to_str(uint8_t *number, uint8_t number_length, char *str_number)
+{
+	/* Copy and log address string */
+	uint8_t length = number_length / 2;
+	bool fill_bits = false;
+
+	if (number_length % 2 == 1) {
+		/* There is one more number in semi-octet and 4 fill bits*/
+		length++;
+		fill_bits = true;
+	}
+
+	uint8_t hex_str_index = 0;
+
+	for (int i = 0; i < length; i++) {
+		/* Handle most significant 4 bits */
+		uint8_t number_value = (number[i] & 0xF0) >> 4;
+
+		if (number_value >= 10) {
+			LOG_WRN("Single number in phone number is higher than 10: index=%d, number_value=%d, lower semi-octet",
```

Comment:
When can this happen?

Reply:
Somebody sends us an SMS PDU which has the hexadecimal formatted phone number with hexa characters of A-F. E.g., 123456789A would be such a phone number. This should never happen but as this is not controlled by our chipset, we must handle it somehow.